src/Symfony/Component/Messenger/Asynchronous/Routing/AbstractSenderLocator.php

```diff
@@ -11,28 +11,33 @@
 
 namespace Symfony\Component\Messenger\Asynchronous\Routing;
 
+use Symfony\Component\Messenger\Envelope;
+
 /**
  * @author Samuel Roze <samuel.roze@gmail.com>
  *
  * @internal
  */
 abstract class AbstractSenderLocator implements SenderLocatorInterface
 {
-    public static function getValueFromMessageRouting(array $mapping, $message)
+    public static function getValueFromMessageRouting(array $mapping, Envelope $envelope)
     {
-        if (isset($mapping[\get_class($message)])) {
-            return $mapping[\get_class($message)];
-        }
-        if ($parentsMapping = array_intersect_key($mapping, class_parents($message))) {
-            return current($parentsMapping);
+        if (isset($mapping[$class = \get_class($envelope->getMessage())])) {
+            return $mapping[$class];
         }
-        if ($interfaceMapping = array_intersect_key($mapping, class_implements($message))) {
-            return current($interfaceMapping);
+
+        foreach (class_parents($class) as $name) {
```

Comment:
@sroze is it a bug that here, we check parents before interfaces, whereas in `AbstractHandlerLocator` we check in the opposite order?

Reply:
AbstractHandlerLocator updated